File: python/project/student_score_manager/student.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import List, Dict


@dataclass
class Student:
    name: str
    scores: List[float] = field(default_factory=list)

    # ---------- score ops ----------
    def add_score(self, value: float) -> None:
        self._validate_score(value)
        # TODO: implement penambahan score

    def edit_score(self, index: int, new_value: float) -> None:
        if index < 0 or index >= len(self.scores):
            raise IndexError("Index nilai tidak valid")
        self._validate_score(new_value)
        # TODO: edit score dengan specific index

    def remove_score(self, index: int) -> None:
        if index < 0 or index >= len(self.scores):
            raise IndexError("Index nilai tidak valid")
        # TODO: hapus score dengan specific index

    # ---------- derived metrics ----------
    def average(self) -> float:
        # TODO: hitung rata-rata score, kembalikan 0.0 jika tidak ada score
        return sum(self.scores) / len(self.scores) if self.scores else 0.0

    def grade(self) -> str:
        avg = self.average()
        # TODO: kembalikan grade berdasarkan kondisi average
        return "E"

    # ---------- utils ----------
    # TODO: buat fungsi infoLine() yang return string dengan format:
    # Nama | scores=[..] | avg=.. | grade=..

    def to_dict(self) -> Dict:
        return {"name": self.name, "scores": self.scores}

    @staticmethod
    def from_dict(d: Dict) -> "Student":
        return Student(name=d["name"], scores=[float(x) for x in d.get("scores", [])])

    @staticmethod
    def _validate_score(value: float) -> None:
        try:
            v = float(value)
        except Exception as e:
            raise ValueError("Nilai harus berupa angka") from e
        if v < 0 or v > 100:
            raise ValueError("Score harus 0..100")

File: java/project/student_score_manager/Student.java
import java.util.ArrayList;
import java.util.List;

public class Student {
    private String name;
    private ArrayList<Double> scores;

    public Student(String name) {
        this.name = name;
        this.scores = new ArrayList<>();
    }

    public String getName() { return name; }
    public void setName(String newName) { this.name = newName; }

    public List<Double> getScores() { return scores; }

    public void addScore(double score) {
        // TODO: add score dengan validasi 0..100
    }

    public void editScore(int index, double newScore) {
        if (index < 0 || index >= scores.size()) throw new IndexOutOfBoundsException("Index nilai tidak valid");
        if (newScore < 0 || newScore > 100) throw new IllegalArgumentException("Score harus 0..100");
        // TODO: edit score dengan specific index
    }

    public void removeScore(int index) {
        if (index < 0 || index >= scores.size()) throw new IndexOutOfBoundsException("Index nilai tidak valid");
        // TODO: hapus score dengan specific index
    }

    public double getAverage() {
        // TODO: hitung rata-rata dari scores ArrayList, kembalikan 0 jika belum ada nilai
        return 0.0;
    }

    public String getGrade() {
        double avg = getAverage();
        // TODO: kembalikan grade berdasarkan kondisi average
        return "E";
    }

    // TODO: buat fungsi infoLine() yang return string dengan format:
    // Nama | scores=[..] | avg=.. | grade=..
}
